clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/JsonTemplatesProducer.kt

```diff
@@ -0,0 +1,252 @@
+package com.clevertap.android.sdk.inapp.customtemplates
+
+import com.clevertap.android.sdk.CleverTapInstanceConfig
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.ACTION
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.BOOLEAN
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.FILE
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.NUMBER
+import com.clevertap.android.sdk.inapp.customtemplates.TemplateArgumentType.STRING
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * A [TemplateProducer] that creates templates based on a json definition. Invalid definitions will
+ * throw [CustomTemplateException] when [defineTemplates] is called.
+ *
+ * @param jsonTemplatesDefinition A string with a json definition of templates in the following format:
+ *```
+ *{
+ *  "TemplateName": {
+ *    "type": "template",
+ *    "arguments": {
+ *      "Argument1": {
+ *        "type": "string|number|boolean|file|action|object",
+ *        "value": "val" // different type depending on "type", e.g 12.5, true, "str" or {}
+ *        },
+ *      "Argument2": {
+ *        "type": "object",
+ *        "value": {
+ *          "Nested1": {
+ *            "type": "string|number|boolean|object", // file and action cannot be nested
+ *            "value": {}
+ *          },
+ *          "Nested2": {
+ *            "type": "string|number|boolean|object",
+ *            "value": "val"
+ *          }
+ *        }
+ *      }
+ *    }
+ *  },
+ *  "functionName": {
+ *    "type": "function",
+ *    "isVisual": true|false,
+ *    "arguments": {
+ *      "a": {
+ *      "type": "string|number|boolean|file|object", // action arguments are not supported for functions
+ *      "value": "val"
+ *      }
+ *    }
+ *  }
+ *}
+ *```
+ *
+ * @param templatesPresenter A presenter for all templates in the json definitions. Required if there
+ * is at least one template with type "template".
+ * @param functionsPresenter A presenter for all functions in the json definitions. Required if there
+ * is at least one template with type "function".
+ */
+open class JsonTemplatesProducer(
+    private val jsonTemplatesDefinition: String,
+    private val templatesPresenter: TemplatePresenter?,
+    private val functionsPresenter: FunctionPresenter?
+) : TemplateProducer {
+
+    override fun defineTemplates(ctConfig: CleverTapInstanceConfig): Set<CustomTemplate> {
```

Comment:
Seems like `ctConfig: CleverTapInstanceConfig` is unused

Reply:
I have let `JsonTemplateProducer` be an open class. If clients need to use `ctConfig` to differentiate between instances, they could extend this class, override `defineTemplates` and use `ctConfig` to check whether to call the super implementation or not.